### PR TITLE
Colors refactor

### PR DESCRIPTION
**Pull request by OzoneH3**
Preparation for changeable colors.
- Colors stored in unordered_map.

```
struct stColors {
    nc_color color; // Default color
    std::string sCustom; // Custom color name
    std::string sInvert; // Invert
    std::string sNoBrigt; // No Bright
};

std::unordered_map<std::string, stColors> mapColors;
```
- Moved enum colornames to defines.

```
#define c_black all_colors.get("c_black")
```
- All colors are defined in load_default();
- Dynamically find < color >_background() colors.
- Probably screws up saved artefact colors.

### CODE REVIEW

File: src/iuse_software_kitten.cpp [hunk 1]
```diff
@@ -295,7 +298,10 @@ robot_finds_kitten::robot_finds_kitten(WINDOW *w)
         do {
             bogus[c].character = ktile[rand() % 82];
         } while (bogus[c].character == '#' || bogus[c].character == ' ');
-        bogus[c].color = int_to_color((rand() % (maxcolor - 2)) + 2);
+
+        do {
+            bogus[c].color = all_colors.get_random();
+        } while ( all_colors.get(kitten.color) == "c_black" );
```
Comment (BevapDin) [inline src/iuse_software_kitten.cpp]: 1. You meant to test `bogus[c].color`.
2. How about a function `all_colors.get_random_non_black()` or similar?

File: src/color.h [hunk 2]
```diff
@@ -181,10 +210,10 @@ nc_color magenta_background(nc_color c);
...
 nc_color rand_color();
 
-nc_color color_from_string(std::string const &color);
-std::string const& string_from_color(nc_color color);
-nc_color bgcolor_from_string(std::string const &color);
-nc_color get_color_from_tag(const std::string &s, nc_color base_color);
+nc_color color_from_string(const std::string  &color);
+std::string const& string_from_color(const nc_color &color);
+nc_color bgcolor_from_string(const std::string  &color);
+nc_color get_color_from_tag(const std::string &s, const nc_color base_color);
```
Comment (BevapDin) [inline src/color.h]: I think these changes are pointless.

The `const` in `const nc_color base_color` is definitely pointless, see http://stackoverflow.com/questions/17208570/top-level-const-doesnt-influence-a-function-signature, it's ignored here anyway

`void f( int x );` and `void f( const int );` define exactly the same function, they are not overloaded functions.

File: src/color.h [hunk 3]
```diff
@@ -15,161 +170,35 @@ HI = 1,
...
- c_cyan_cyan  = COLOR_PAIR(71),
- c_ltcyan_cyan  = COLOR_PAIR(71) | A_BOLD
+typedef int nc_color;
 
+class clColors {
+    public:
+        clColors() {};
+
+        nc_color get(const std::string &sName);
+        std::string get(const nc_color &color);
```
Comment (BevapDin) [inline src/color.h]: Taking a const reference is only useful if the object is big or copying it is expensive. Neither is true for an int (which nc_color is). I recommend therefor to take the nc_color parameters as by-value, through a const reference is not incorrect.

File: src/color.cpp [hunk 4]
```diff
@@ -775,200 +431,51 @@ nc_color rand_color()
...
  * @param new_color The color to get, as a std::string.
  * @return The nc_color constant that matches the input.
  */
-nc_color color_from_string(std::string const &new_color)
+nc_color color_from_string(const std::string &color)
 {
-    auto const it = colormap.find(new_color);
-    if (it != std::end(colormap)) {
-        return it->second;
+    std::string new_color = color;
```
Comment (BevapDin) [inline src/color.cpp]: If you're going to copy the parameter anyway, why not let the caller do the copying (take the parameter by value)? This also avoids a copy in certain situations.

File: src/color.cpp [hunk 5]
```diff
@@ -6,756 +6,412 @@
...
- init_pair(67, COLOR_GREEN, COLOR_CYAN);
- init_pair(68, COLOR_YELLOW, COLOR_CYAN);
- init_pair(69, COLOR_BLUE, COLOR_CYAN);
- init_pair(70, COLOR_MAGENTA, COLOR_CYAN);
- init_pair(71, COLOR_CYAN, COLOR_CYAN);
+    if ( mapColors[sName].first == 0 ) {
+        return 0;
+    }
 
+    return ( mapColors[sName].second != "" ) ? mapColors[mapColors[sName].second].first : mapColors[sName].first;
```
Comment (BevapDin) [inline src/color.cpp]: The map look up `mapColors[sName]` is done 3 times, why not use an iterator and search the entry only once?
```C++
auto const iter = mapColors.find( sName );
if( iter == mapColors.end() ) {
    // not found -> return default or whatever
    return ...;
}
auto &entry = iter->second; // this is the same as mapColors[sName]
if( entry.first == 0 ) { return 0; }
....
```

File: src/color.cpp [hunk 6]
```diff
@@ -6,756 +6,412 @@
...
-   case c_ltblue_cyan  : return 139;
-   case c_magenta_cyan  : return 140;
-   case c_pink_cyan  : return 141;
-   case c_cyan_cyan  : return 142;
-   case c_ltcyan_cyan  : return 143;
-
- }
- return 0;
+    for ( const auto& iter : mapColors ) {
+        if ( iter.second.first == color ) {
```
Comment (BevapDin) [inline src/color.cpp]: At this point, nobody knows what `second`, `first` or what `second.first` (third?). is How about making a small struct that gives those values meaningful names:
```C++
struct invert_defs {
    std::string color_invert;
    std::string color_bright;
};
struct custom_defs {
    nc_color color_default;
    std::string color_custom;
};
```

You might know a better name, I'm not creative with naming.

File: src/color.cpp [hunk 7]
```diff
@@ -6,756 +6,412 @@
...
-   case 140: return c_magenta_cyan;
-   case 141: return c_pink_cyan;
-   case 142: return c_cyan_cyan;
-   case 143: return c_ltcyan_cyan;
-
- }
- return c_black;
+    const std::string sName = all_colors.get(color);
+
+    if ( mapInvert[sName].first == "" ) {
```
Comment (BevapDin) [inline src/color.cpp]: Btw. `std::string` has a function `empty()` which does this test, is shorter to read and write and probably a tiny bit faster.

File: src/color.cpp (hunk 4 above)
Comment (OzoneH3) [inline src/color.cpp]: In this case it's used later in the debugmsg(). Changed it in the other functions though.

**Comment by OzoneH3** [comment]
@BevapDin Thanks for the input. Fixed all the things.
I'll leave the get_random() as it is for now, it's only uses in the kitten game atm.

**Pull request merged**